Site Settings: Introduce a Sitemaps card in the Traffic section

tyxla opened the pull request:
### Purpose

This PR introduces the Sitemaps card in the Traffic section, and removes the Sitemap section from the Site Verification Services card. UI of the card has been borrowed from Jetpack. 

Part of #12845. Moving this card above the Site Verification Services card will be done once we separate Site Verification Services from the main form (currently they're all one large component). Or the other way around will be to insert that component in the main SEO form.

### How it works

This card looks and behaves differently for WordPress.com and Jetpack sites. For WordPress.com sites it simply lists the sitemap links, while for Jetpack sites it also allows the user to toggle the Sitemaps module. See previews for more details.

### Previews

**Site Verification Services card - XML sitemaps section is now removed**
![](https://cldup.com/rC7K_kcrjM.png)

**Sitemaps card for WordPress.com sites**
![](https://cldup.com/2fJ9LuSlNA.png)

**Sitemaps card for Jetpack sites, when Sitemaps module is disabled**
![](https://cldup.com/8zzLHIvcPo.png)

**Sitemaps card for Jetpack sites, while Sitemaps module is being activated**
![](https://cldup.com/a4o6x9o1c4.png)

**Sitemaps card for Jetpack sites, when Sitemaps module is active**
![](https://cldup.com/e3MCc2tYtA.png)

### To test
* Checkout this branch, or get it going on calypso.live
* Go to `/settings/traffic/$site`, where `$site` is one of your WordPress.com sites.
* Verify the Sitemaps card is shown at the bottom and it looks as shown on the preview.
* Verify the links in the Sitemaps card look and work right.
* Go to `/settings/traffic/$site`, where `$site` is one of your Jetpack sites.
* Verify the Sitemaps card is shown at the bottom and it looks as shown on the preview.
* Verify the module toggle works correctly as shown on the preview.
* Verify the sitemap links look and work right.

a8ci18n commented on client/my-sites/site-settings/sitemaps.jsx:
```diff
@@ -0,0 +1,196 @@
+			<div>
+				<QueryJetpackConnection siteId={ siteId } />
+
+				<SectionHeader label={ translate( 'Sitemaps' ) } />
```
Hi! I've found a possible matching string that has already been translated **25** times:
`translate( 'Sitemap' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

tyxla replied:
A good suggestion, but we're intentionally going with the plural one.

a8ci18n commented on client/my-sites/site-settings/sitemaps.jsx:
```diff
@@ -0,0 +1,196 @@
+				{
+					activatingSitemapsModule &&
+					<FormSettingExplanation>
+						{ translate( 'Generating sitemap…' ) }
```
Hi! I've found a possible matching string that has already been translated **32** times:
`translate( 'Generating...' )` *ES Score: 8*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/site-settings-sitemaps-card)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

tyxla replied:
Good one, too. Using a different one for consistency with Jetpack.

a8ci18n commented on client/my-sites/site-settings/sitemaps.jsx:
```diff
@@ -0,0 +1,196 @@
+
+		return (
+			<FormSettingExplanation>
+				{ translate( 'Your sitemap is automatically sent to all major search engines for indexing.' ) }
```
Hi! I've found a possible matching string that has already been translated **22** times:
`translate( 'Your site's sitemap is automatically sent to all major search engines for indexing.' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

tyxla replied:
I've stuck with the other one for consistency with Jetpack, but I'm not sure if we should switch to this suggested one - it was used in Calypso in the Site Verification Services card. @rickybanister thoughts?

a8ci18n commented on client/my-sites/site-settings/sitemaps.jsx:
```diff
@@ -0,0 +1,196 @@
+			<div className="sitemaps__info-link-container site-settings__info-link-container">
+				<InfoPopover position={ 'left' }>
+					<ExternalLink href={ url } icon target="_blank">
+						{ translate( 'Learn more about Sitemaps.' ) }
```
Hi! I've found a possible matching string that has already been translated **32** times:
`translate( 'Learn more about CSS' )` *ES Score: 6*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

drw158 commented:
I'm not too involved with all of the settings work, but:

On WordPress.com sites...

- Seems weird that the description is now above the xml link.
- "XML" has been removed from the title, which I think is not as clear if users are looking for "XML Sitemap".
- Can you have 2 sitemaps on wpcom?

Here is mine right now:

<img width="738" alt="screen shot 2017-04-07 at 10 20 05 am" src="https://cloud.githubusercontent.com/assets/618551/24806962/1eef6a3e-1b7c-11e7-8834-a224918a8cd0.png">

rickybanister commented:
@drw158 I don't think it's that weird to have the description above the link. I'd rather learn about what this thing is before clicking that link and being very confused by the contents of it. 

We could add XML back into the title— @MichaelArestad did we remove that just to make it more friendly? (On the Jetpack side)

@tyxla I have one question for you—what do we do here (or what should we do here) if you set your site to private? Should we attach a notice to this card?

tyxla commented:
> "XML" has been removed from the title, which I think is not as clear if users are looking for "XML Sitemap".

I'm fine with renaming it to **XML Sitemaps** or even **XML Sitemap** if that would make sense. Let's see what mister @MichaelArestad would suggest on that one.

> Can you have 2 sitemaps on wpcom?

@drw158 sure thing. All WordPress.com sites have a news sitemap enabled by default, in addition to the default sitemap. Here is your sitemaps, for example: 
* Default: https://davewhitley.wordpress.com/sitemap.xml
* News: https://davewhitley.wordpress.com/news-sitemap.xml

> @tyxla I have one question for you—what do we do here (or what should we do here) if you set your site to private? Should we attach a notice to this card?

Ah, that's a good call @rickybanister. Maybe we could stick with a consistent approach with how it's done in Jetpack right now. I like the current approach there, where it changes the message if your site is non-public:

![](https://cldup.com/YgcCrzvvCS.png)

That one can be refactored as a notice if that makes more sense to you. And we could change its wording a bit, so it would direct to the Calypso setting for managing site visibility. What do you think?

rickybanister commented:
The thing is, I don't think we should let users toggle the feature on if they have their site set to private. I realize that historically the toggle corresponds to the module being activated, but in this case the toggle refers directly to whether a search engine can crawl a sitemap (which it can't). 

Let's go with your text there, but make the toggle disabled if the site is set to private.

Also, the private/public setting is in /settings/general, not 'Reading' on WordPress.com so let's update that.

a8ci18n commented on client/my-sites/site-settings/sitemaps.jsx:
```diff
@@ -0,0 +1,236 @@
+				{
+					isSitePublic && activatingSitemapsModule &&
+					<FormSettingExplanation>
+						{ translate( 'Generating sitemap…' ) }
```
Hi! I've found a possible matching string that has already been translated **32** times:
`translate( 'Generating...' )` *ES Score: 8*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/site-settings-sitemaps-card)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

tyxla commented:
@rickybanister done - I've updated the toggle to be disabled if the site is not public; also updated it to display that text. See updated previews:

**Non-public Jetpack site**
![](https://cloud.githubusercontent.com/assets/8436925/24974652/8ba925d6-1fcb-11e7-9a0f-ce8b7387292a.png)

**Non-public WordPress.com site**
![wpcom](https://cloud.githubusercontent.com/assets/8436925/24974662/97338144-1fcb-11e7-96bf-eabea4955b40.png)

Feel free to review it again @rickybanister. Also, since I've changed some stuff, another code review would be recommended @ryelle.

rickybanister commented:
That works—my only final thought is why do we have to say 'it is possible that Privacy is not set to 'public'? Can't we accurately tell the site's privacy settings? 

If we can properly detect that the text would be better as:

`Your site is not currently accessible to search engines. You must set your %privacy settings% to 'public'.`

tyxla commented:
@rickybanister yes, we accurately detect that. That's what we use to determine what message to display. So I've updated the text:

![](https://cldup.com/lnlWoqL7xB.png)

Feel free to have another 👀

rickybanister commented:
Perfect, ship it.

ryelle commented on client/my-sites/site-settings/sitemaps.jsx:
```diff
@@ -0,0 +1,236 @@
+				{ ! isSitePublic && this.renderNonPublicExplanation() }
+
+				{
+					isSitePublic && activatingSitemapsModule &&
```
Rather than continuing to check isSitePublic, maybe return early with just the nonPublicExplanation?

tyxla replied:
Good idea! Fixed.

ryelle commented:
I'm having some trouble testing this now, and I'm not sure if it's my setup or not. I have a private site, shows up as "hidden" in settings/general, but I'm only seeing a disabled toggle for the sitemaps, no explanation. I reloaded, and got the explanation once, but reloaded again and am back to a disabled toggle. I see this in console, but I'm not sure where it's coming from:

<img width="734" alt="screen shot 2017-04-13 at 3 11 36 pm" src="https://cloud.githubusercontent.com/assets/541093/25020231/8b214040-205b-11e7-8552-3615821c0e83.png">

tyxla commented:
The console errors are coming from `draft.js`, which is used for the SEO settings, and is not related to the PR. These errors have been there for a while 😞 

Regarding the issues with reproducing: can you please try reproducing again, and if the issue persists, can you write down the exact steps you follow when occurring the issue? Or a gif could also help a lot. Thanks!

ryelle commented:
Here's a more thorough set of test cases:

- Tested on a public wp.com site, shows the sitemap links ✅
- Set that site to "Hidden", shows the not public explanation ✅
- Public Jetpack site, shows the toggle, I can turn it on and then it shows the sitemap links ✅
- Set _that_ site to "Hidden", and it shows the disabled toggle (on), with the not public explanation ✅
- Set back to public so I could turn off sitemaps, then set back to hidden, and I only see the disabled toggle (off). ❌ 

This is what was happening in my last comment. I would expect to see the not public explanation here too:
<img width="739" alt="screen shot 2017-04-13 at 3 57 16 pm" src="https://cloud.githubusercontent.com/assets/541093/25021937/05785eb8-2062-11e7-82de-13a694f5edd0.png">

tyxla commented:
@ryelle thanks for the review! I've addressed the issue in 4e50b57330acdcbd8e0e4d6f439683a6700eb322. Feel free to give it another try.

sirreal commented:
* ✅  Tested on a public wp.com site, shows the sitemap links.
* ✅  Set that site to "Private", shows the "must set to public" explanation.
* ✅  Set that site to "Hidden", shows the "must set to public" explanation.
* ✅  Public Jetpack site, shows the toggle, I can turn it on and then it shows the sitemap links
* ✅  Set that site to "Hidden", and it shows the disabled toggle (on), with the not public explanation
* ✅  Set back to public so I could turn off sitemaps, then set back to hidden, and I only see the disabled toggle (off).

tyxla commented:
Thanks for testing @sirreal! 

@ryelle can you confirm this resolves the issue for you?

MichaelArestad commented:
The current JP version looks like this:

![image](https://cloud.githubusercontent.com/assets/1123119/25402354/ac0cafca-29b5-11e7-9d86-e44cfa56af08.png)

I don't think most people using this even know or care what XML is. I'm okay with the verbiage either way.

ryelle commented:
Perfect, I see the not-public explanation now, even if sitemaps are disabled. Looks good to 🚢
